fix unflatten_dense_tensor when there is empty tensor inside

zhaojuanmao opened the pull request:
Summary:
Quantization team reported that when there are two empty tensors are replicated among ranks, the two empty tensors start to share storage after resizing.

The root cause is unflatten_dense_tensor unflattened the empty tensor as view of flat tensor and thus share storage with other tensors.

This PR is trying to avoid unflatten the empty tensor as view of flat tensor so that empty tensor will not share storage with other tensors.

Test Plan: unit test

Differential Revision: D25859503

vkuzo commented:
Thanks for fixing this!  Test plan LG, code lg as well but I'd hope to wait from someone who is more familiar with the code to accept.

zhaojuanmao commented:
@mrshenli @pritamdamania87 would you please help reviewing this diff? thanks!

pritamdamania87 commented on torch/csrc/utils/tensor_flatten.h:
```diff
@@ -20,8 +20,17 @@ inline std::vector<at::Tensor> unflatten_dense_tensors(const at::Tensor& flat, a
+    // with other unflatten tensors.
+    auto flat_tensor_view = flat.narrow(0, offset, numel).view(tensor.sizes());
+    if (numel == 0) {
+      outputs.push_back(at::empty({0}, flat_tensor_view.options()));
```
nit: We don't need flat_tensor_view here right? We can just use flat.options()? Then we only need to create `flat_tensor_view` in the else block below?

pritamdamania87 commented on test/cpp/api/tensor_flatten.cpp:
```diff
@@ -0,0 +1,38 @@
+  ASSERT_EQ(unflatten_results.at(3).numel(), 2);
+
+  // empty tensor address is 0 as memory is not allocated yet
+  ASSERT_EQ(unflatten_results.at(1).data_ptr(), unflatten_results.at(2).data_ptr());
```
Should we assert here that each one's data_ptr is 0 instead of asserting both data_ptrs are equal? This assertion seems to suggest both should share the same storage (which they shouldn't). They're equal just from the fact that they are empty for now.

zhaojuanmao replied:
sounds good!

zhaojuanmao committed:
fix unflatten_dense_tensor when there is empty tensor inside (#50321)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/50321

Quantization team reported that when there are two empty tensors are replicated among ranks, the two empty tensors start to share storage after resizing.

The root cause is unflatten_dense_tensor unflattened the empty tensor as view of flat tensor and thus share storage with other tensors.

This PR is trying to avoid unflatten the empty tensor as view of flat tensor so that empty tensor will not share storage with other tensors.

Test Plan: unit test

Reviewed By: pritamdamania87

Differential Revision: D25859503

fbshipit-source-id: 4cb9756911a3bcaf46daf66b5a66b08fe9f1bda5
